postgres-protocol/src/misc/mod.rs

```diff
@@ -0,0 +1,92 @@
+//! As always, there are some functions that just don't fit anywhere.
+
+use hmac::{Hmac, Mac, NewMac};
+use rand::RngCore;
+use sha2::digest::FixedOutput;
+use sha2::{Digest, Sha256};
+
+#[cfg(test)]
+mod test;
+
+const SCRAM_DEFAULT_ITERATIONS: u32 = 4096;
+const SCRAM_DEFAULT_SALT_LEN: usize = 16;
+
+/// Choose the password encryption algorithm.
+#[derive(Debug)]
+pub enum PasswordAlgorithm {
+    /// Avoid `MD5`, which is not considered to be secure.
+    MD5,
+    /// [SCRAM-SHA-256](https://en.wikipedia.org/wiki/Salted_Challenge_Response_Authentication_Mechanism)
+    ScramSha256,
+}
+
+/// Error encrypting password.
+#[derive(Debug)]
+pub struct EncryptionError;
+
+/// Encrypt a password in the client.
+///
+/// This is intended to be used by client applications that wish to
+/// send commands like ALTER USER joe PASSWORD 'pwd'. The password
+/// need not be sent in cleartext if it is encrypted on the client
+/// side. This is good because it ensures the cleartext password won't
+/// end up in logs pg_stat displays, etc.
+///
+/// The client may assume the returned string doesn't contain any
+/// special characters that would require escaping.
+pub fn encrypt_password(
+    passwd: &str,
+    user: &str,
+    algorithm: PasswordAlgorithm,
+) -> Result<String, EncryptionError> {
+    match algorithm {
+        PasswordAlgorithm::MD5 => {
+            let crypt = format!("{}{}", passwd, user);
+            let digest = md5::compute(crypt.as_bytes());
+            Ok(format!("md5{:x}", digest))
+        }
+        PasswordAlgorithm::ScramSha256 => {
+            let prepared_passwd = stringprep::saslprep(passwd)
+                .map_err(|_| EncryptionError)?
+                .into_owned();
+            Ok(scram_sha_256(prepared_passwd.into_bytes().as_slice()))
+        }
+    }
+}
+
+fn scram_sha_256(prepared_passwd: &[u8]) -> String {
```

Comment:
Yeah, we should be able to deduplicate this code with the SASL implementation somehow I think.